docs/src/tutorials/hybrid-applications-with-webpack/template-deep-dive/src-lib-rs.md

```diff
@@ -0,0 +1,306 @@
+`#[wasm_bindgen]` attribute, then `run` would not be easily
+accessible within JavaScript. Furthermore, we wouldn't be
+able to natively convert certain types such as `()` between
+JavaScript and Rust. So, the `#[wasm_bindgen]` attribute
+allows `run` to be called from JavaScript.
+
+This is all you need to know to interface with JavaScript!
+If you are curious about the rest, read on.
+
+## 2. Crate imports
+
+```rust
+#[macro_use]
+extern crate cfg_if;
+extern crate web_sys;
+extern crate wasm_bindgen;
+```
+
+In `Cargo.toml`, we included the crates `cfg_if`, `web_sys`,
+and `wasm_bindgen` as project dependencies.
+
+Here, we explicitly declare that these crates will be used
+in `lib.rs`.
+
+```rust
+use wasm_bindgen::prelude::*;
+```
+
+`use` allows us to conveniently refer to parts of a crate or
+module. For example, suppose the crate `wasm_bindgen`
+contains a function `func`. It is always possible to call
+this function directly by writing `wasm_bindgen::func()`.
+However, this is often tedious to write. If we first specify
+`use wasm_bindgen::func;`, then `func` can be called by just
+writing `func()` instead.
+
+In our `use` statement above we further specify a `prelude`
+module. Many modules contain a "prelude", a list of things
+that should be automatically imported. This allows common
+features of the module to be conveniently accessed without a
+lengthy prefix. For example, in this file we can use
+`#[wasm_bindgen]` only because it is brought into scope by
+the prelude.
+
+The asterisk at the end of this `use` indicates that
+everything inside the module `wasm_bindgen::prelude` (i.e.
+the module `prelude` inside the crate `wasm_bindgen`) can be
+referred to without prefixing it with
+`wasm_bindgen::prelude`.
+
+For example, `#[wasm_bindgen]` could also be written as
+`#[wasm_bindgen::prelude::wasm_bindgen]`, although this is
+not recommended.
+
+One other point of interest is how we import the `cfg_if!`
+macro.
+
+```rust
+#[macro_use]
+extern crate cfg_if;
+```
+
+The `#[macro_use]` attribute imports the `cfg_if!` macro the
+same way a `use` statement imports functions.
+
+
+## 3. `wee_alloc` optional dependency
+
+```rust
+cfg_if! {
+    // When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
+    // allocator.
+    if #[cfg(feature = "wee_alloc")] {
+        extern crate wee_alloc;
+        #[global_allocator]
+        static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+    }
+}
+```
+
+This code block is intended to initialize `wee_alloc` as the
+global memory allocator, but only if the `wee_alloc` feature
+is enabled in `Cargo.toml`.
+
+We immediately notice that `cfg_if!` is a macro because it
+ends in `!`, similarly to other Rust macros such as
+`println!` and `vec!`. A macro is directly replaced by other
+code during compile time.
+
+During compile time, `cfg_if!` evaluates the `if` statement.
+This tests whether the feature `wee_alloc` is present in the
```

Comment:
Technically the feature is always present in the manifest, but it's rather checking to see if it was enabled at compiled time (e.g. via `--features`)